### ISSUE TITLE
No Anytone 578 Support

### ISSUE THREAD

**Issue by meleschi** [sergev/dmrconfig#53]
Good morning,

It looks like there is no support for the Anytone 578.  I regularly use your product to update my 878, and now own a 578.  Can I help in any way to get 578 support into DMRConfig?

**Comment by nicolasjon** [comment]
Unfortunately the firmware and CPS keeps changing which does not help.
Sadly it seems this project has been abandoned. As no response to emails or pull requests. Which is a great shame.
I am considering producing a fork at least for the 868/878 series as the work done so far is of high quality and is a great alternative to the OEM CPS which will only ever be Windows GUI. Thoughts?

**Comment by jasiek** [comment]
100% do a fork.

**Comment by jpfox** [comment]
Anybody know what to do to use dmrconfig with a at-d578UV?

**Comment by hmatuschek** [comment]
There has been very little activity here for a while. However, @reald and myself are interested in reverse engineering the 578 codeplug and consequently implementing it in qdmr. Fortunately, it is possible to emulate these devices in software and tricking the CPS to send the codeplug to the emulator. To do that, however, we need to know how the device identifies itself to the CPS. So it would help enormously if someone could capture one codeplug write to the device using wireshark (https://www.wireshark.org/) and get us the pcapng file. The content of the codeplug does not matter, as we only need to know how the device identifies itself. Hence you may simply write the default codplug.

**Comment by wspeirs** [comment]
@hmatuschek I fired up my D578UV III Plus and did a read from the radio, then wrote the same plug back to the radio. I'm new to this radio, and new to USB protocols, but hopefully this capture helps get your question answered.

That said, I'm a pretty experienced programmer (not as much w/embedded, but would love to learn!), and would love to get this working for the D578. How can I help?

[D587UVIII_Plus_read_from_radio.pcapng.gz](https://github.com/OpenRTX/dmrconfig/files/13245811/D587UVIII_Plus_read_from_radio.pcapng.gz)

[D587UVIII_Plus_write_to_radio.pcapng.gz](https://github.com/OpenRTX/dmrconfig/files/13245825/D587UVIII_Plus_write_to_radio.pcapng.gz)

**Comment by wspeirs** [comment]
OK, did some quick hacking, and was able to read from my D578UV III Plus; see https://github.com/wspeirs/dmrconfig

My $0.02 would be to close other issues for the D578 and focus efforts on this issue

**Comment by jsaern** [comment]
How correct is it to treat the D578UV III Plus as "identical" to a D868?

How risky is just reading the config out? One of the articles in the documentation (the Danger Zone) suggested that even reading out from the radio would harm the radio.

**Comment by jsaern** [comment]
Found the information I needed: https://kd9cpb.com/qdmr#comment-489

TL;DR
Reading and decoding likely harmless. 

NOTE
Reading radio as if it is a D878 seems preferred over what WSpeirs observed.

**Comment by jsaern** [comment]
I read what little information I had managed to program into my D578UVIII using the hardware buttons.  Decoded it and found the errors below. 

(I also read the radio using QDMR and a cursory look suggested it looked OK. Need to examine closer.)

How likely these errors are from incomplete programming using the buttons?

```
dmrconf decode --radio=d878uv codeplug-578.dfu codepl
ug-decoded.conf
```
```
Warning in lib/anytone_codeplug.cc@617: Mixed mode channels are not supported (for now). Treat ch 'Channel 4' as digital channel.
Warning in lib/anytone_codeplug.cc@588: Mixed mode channels are not supported (for now). Treat ch 'Channel 7' as analog channel.
ERROR in lib/d878uv_codeplug.cc@2865: Cannot link AnyTone settings extension: second auto-repeater VHF offset index 2 not defined.
ERROR in lib/d878uv_codeplug.cc@4230: Cannot link general settings extension.
ERROR in lib/anytone_codeplug.cc@4530: Cannot decode AnyTone codeplug: Linking of config objects failed.
ERROR in cli/decodecodeplug.cc@53: Cannot decode binary codeplug file 'codeplug-578.dfu'.
ERROR in cli/decodecodeplug.cc@150: Cannot decode codeplug 'codeplug-578.dfu':
  In decodecodeplug.cc:53: Cannot decode binary codeplug file 'codeplug-578.dfu'.
  In anytone_codeplug.cc:4530: Cannot decode AnyTone codeplug: Linking of config objects failed.
  In d878uv_codeplug.cc:4230: Cannot link general settings extension.
  In d878uv_codeplug.cc:2865: Cannot link AnyTone settings extension: second auto-repeater VHF offset index 2 not defined.
```

**Comment by jsaern** [comment]
How much would it help if the BCDMR volunteers programmed my radio and I collected the error messages from reading and decoding that?

https://bcdmr.wordpress.com/2021/12/10/is-there-a-code-plug-for-the-d578uviii-pro-mobile/

**Comment by hmatuschek** [comment]
> How correct is it to treat the D578UV III Plus as "identical" to a D868?
> 
> How risky is just reading the config out? One of the articles in the documentation (the Danger Zone) suggested that even reading out from the radio would harm the radio.

No, it won't. Writing however, may brick it.

**Comment by hmatuschek** [comment]
> I read what little information I had managed to program into my D578UVIII using the hardware buttons. Decoded it and found the errors below.
> 
> (I also read the radio using QDMR and a cursory look suggested it looked OK. Need to examine closer.)
> 
> How likely these errors are from incomplete programming using the buttons?
> 
> ```
> dmrconf decode --radio=d878uv codeplug-578.dfu codepl
> ug-decoded.conf
> ```
> 
> ```
> Warning in lib/anytone_codeplug.cc@617: Mixed mode channels are not supported (for now). Treat ch 'Channel 4' as digital channel.
> Warning in lib/anytone_codeplug.cc@588: Mixed mode channels are not supported (for now). Treat ch 'Channel 7' as analog channel.
> ERROR in lib/d878uv_codeplug.cc@2865: Cannot link AnyTone settings extension: second auto-repeater VHF offset index 2 not defined.
> ERROR in lib/d878uv_codeplug.cc@4230: Cannot link general settings extension.
> ERROR in lib/anytone_codeplug.cc@4530: Cannot decode AnyTone codeplug: Linking of config objects failed.
> ERROR in cli/decodecodeplug.cc@53: Cannot decode binary codeplug file 'codeplug-578.dfu'.
> ERROR in cli/decodecodeplug.cc@150: Cannot decode codeplug 'codeplug-578.dfu':
>   In decodecodeplug.cc:53: Cannot decode binary codeplug file 'codeplug-578.dfu'.
>   In anytone_codeplug.cc:4530: Cannot decode AnyTone codeplug: Linking of config objects failed.
>   In d878uv_codeplug.cc:4230: Cannot link general settings extension.
>   In d878uv_codeplug.cc:2865: Cannot link AnyTone settings extension: second auto-repeater VHF offset index 2 not defined.
> ```

One quick comment. dmrconf is the command line tool from qdmr. This project is dmrconfig.

**Comment by jsaern** [comment]
Sorry, Hannes, I reached this discussion in my latest round of checking on DMRconfig and QDMR. 

I just returned from having VE7CQT program the radio using the official CPS. 

I have both DMRconfig and QDMR and intend to try both for reading the data.  Hopefully, what I find out will help.

**Comment by jsaern** [comment]
`dmrconfig -r -t`

```
Cannot find USB device 0483:df11
Cannot find USB device 15a2:0073
Cannot find USB device 28e9:018a
No radio detected.
Check your USB cable!
```

Anytone AT-D578UV III Plus (Version 2)  
DMR Config, Version 1.1.258
